lib/geolookup/usa/areacodes.rb

```diff
@@ -3,14 +3,103 @@ module USA
     module AreaCodes
 
       AREA_CODE_META_INFO_FILE = 'AREA_CODE_META_INFO.yml'
+      AREACODE_POPULATION_FILE = 'CITY_STATE_COUNTY_POPULATION.csv'
 
       def self.to_h
         @area_codes_hash ||= Geolookup.load_hash_from_file(AREA_CODE_META_INFO_FILE)
       end
 
+      def self.details
+        @records ||= Geolookup.load_list_from_csv(AREACODE_POPULATION_FILE).map do |r|
+          Detail.new(areacode: r[0].to_i,
+                     city: r[1],
+                     statecode: r[2].to_i,
+                     countycode: r[3].to_i,
+                     population: r[4].to_i)
+        end
+      end
+
       def self.find(area_code)
         to_h[area_code.to_s] || {}
       end
+
+      # Return a list of detail records for areacode,
+      # sorted by population in descending order.
+      def self.major_cities(areacode)
+        @major_cities_by_areacode ||= create_major_cities_by_areacode
+        @major_cities_by_areacode.fetch(areacode.to_i, [])
+      end
+
+      # Return a list of areacodes for city and statecode.
+      def self.areacodes_by_city_and_statecode(city, statecode)
+        @areacodes_by_city_and_state ||= create_areacodes_by_city_and_statecode
+        @areacodes_by_city_and_state.fetch(statecode).fetch(city.downcase, [])
+      end
+
+      # Return a list of areacodes for countycode and statecode
+      def self.areacodes_by_countycode_and_statecode(countycode, statecode)
+        @areacodes_by_county_and_state ||= create_areacodes_by_countycode_and_statecode
+        @areacodes_by_county_and_state.fetch(statecode).fetch(countycode, [])
+      end
+
+      private
+
+      def self.create_areacodes_by_countycode_and_statecode
+        h = {}
+        details.group_by{ |r| r.statecode }.each do |state, records|
+          counties = {}
+          h[state] = counties
+          records.group_by{ |r| r.countycode }.each do |county, countyrecords|
+            counties[county] = countyrecords.map{ |r| r.areacode }.sort.uniq
+          end
+        end
+        h
+      end
```

Comment:
if you wanna make these actually private, you can use `private_class_method`
```ruby
def self.some_meth
end; private_class_method :some_meth
```

Reply:
what in the blue hell

Reply:
That kinds looks uglier and difficult to maintain. I just added "private" as a visual cue that everything below that is not accessed by other classes. Should I leave it as is?

Reply:
Yeah doesn't really matter for this, not much could be done by someone reaching into there